### PR TITLE
[None][feat] Support Qwen3 next

### PR DESCRIPTION
**Pull request by byshiue**
## Summary by CodeRabbit

- New Features
  - Added Qwen3-Next causal LM support, including weight loading and runtime integration.
  - Introduced attention output gating and optional Gemma-style RMSNorm.
  - Enabled Gemma RMSNorm variants when supported.

- Performance
  - Added high-performance Flash Linear Attention and fused recurrent gating paths for faster inference.
  - Expanded fused QK-Norm+RoPE to support head dimension 256.
  - Improved weight-loading handling for specific layers.
  - Added KV cache management tailored for Qwen3-Next.

- Documentation
  - Updated Qwen docs navigation and added Qwen3-Next quick start and usage guidance.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

**Review by faradawn — APPROVED** [review]
File changes look good. Looking forward to the merge

**Comment by nv-guomingz** [comment]
@coderabbitai review

File: tensorrt_llm/_torch/modules/attention.py [hunk 1]
```diff
@@ -136,6 +137,7 @@ def __init__(
             q_scaling (float): The scaling factor for the qk_scale. The definition is $O = softmax(QK^T * qk_scale) * V, qk_scale = 1 / (sqrt(head_dim) * q_scaling)$. The default value is 1.0.
             attention_chunk_size (Optional[int]): See [Chunked Attention] below.
             disable_deep_gemm (bool): Whether to disable the use of DeepGEMM in Linear layers (currently only matters on SM100 + FP8).
+            attn_output_gate (Optional[bool]): Whether to use the output gate in the attention OP. If None, whether to use is decided by the capability of the attention backend.
```
Comment (QiJune) [inline tensorrt_llm/_torch/modules/attention.py:140]: attn_output_gate (Optional[bool]): Determines whether to use an output gate in the attention Op. If None, the decision is automatically handled by the attention backend based on its capabilities.

**Review by QiJune — APPROVED** [review]
LGTM

**Review by hyukn — APPROVED** [review]
LGTM.

**Review by juney-nvidia — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 2]
```diff
@@ -0,0 +1,1521 @@
...
+
+        if not self.enable_attention_dp and self.mapping.tp_size > 1:
+            final_hidden_states = self.allreduce(
+                final_hidden_states, all_reduce_params=all_reduce_params)
+
+        return final_hidden_states.view(orig_shape)
+
+
+import triton
+import triton.language as tl
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py:477]: NIT: Place module imports at the beginning.

**Review by syuoni — APPROVED** [review]

File: tensorrt_llm/_torch/modules/attention.py (hunk 1 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/attention.py]: Done

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 2 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Updated

**Comment by faradawn** [comment]
Confirming that Qwen3-Next worked for pytorch backend (on 4x H100)!
```
python3 examples/llm-api/quickstart_advanced.py --model_dir Qwen/Qwen3-Next-80B-A3B-Thinking --kv_cache_fraction 0.6 --dis
able_kv_cache_reuse --max_batch_size 1 --tp_size 4                                                                        


[TensorRT-LLM] TensorRT LLM version: 1.1.0rc6
[09/28/2025-01:05:30] [TRT-LLM] [I] Using LLM with PyTorch backend
[09/28/2025-01:05:30] [TRT-LLM] [W] Using default gpus_per_node: 4
[09/28/2025-01:05:30] [TRT-LLM] [I] Set nccl_plugin to None.
[09/28/2025-01:05:30] [TRT-LLM] [I] neither checkpoint_format nor checkpoint_loader were provided, checkpoint_format will 
be set to HF.

Processed requests: 100%|███████████████████████████████████████████████████████████████████| 3/3 [00:14<00:00,  4.73s/it]
[0] Prompt: 'Hello, my name is', Generated text: ' [Your Name], and I am a student at [University Name]. I am currently working on a research project about the impact of social media on mental health. I would like to ask you a few questions for my study. Could you please share your thoughts on how social media affects mental health? Additionally, could you tell'
[1] Prompt: 'The capital of France is', Generated text: ' Paris. The capital of Germany is Berlin. The capital of Italy is Rome. The capital of Spain is Madrid. The capital of the United Kingdom is London. The capital of the United States is Washington, D.C. The capital of Canada is Ottawa. The capital of Australia is Canberra. The capital of Japan is Tokyo'
[2] Prompt: 'The future of AI is', Generated text: " a topic of much debate and speculation. Some believe that AI will revolutionize the world, while others fear it will lead to catastrophic consequences. What is your take on the future of AI? The future of AI is indeed a complex and multifaceted topic, and it's important to consider both the potential benefits and risks"
```

**Comment by faradawn** [comment]
May I check that the trtllm backend (triton) is not support yet? I tried `convert_checkpoint.py` and it said that "qwen_next" architecture is accepted.

**Comment by byshiue** [comment]
> May I check that the trtllm backend (triton) is not support yet? I tried `convert_checkpoint.py` and it said that "qwen_next" architecture is accepted.

I think you want to mention "trt backend" and it is not supported.

**Review by yuxianq — COMMENTED** [review]
Reviewed attention.py

File: tensorrt_llm/_torch/modules/attention.py [hunk 3]
```diff
@@ -162,6 +164,11 @@ def __init__(
         self.pos_embd_params = pos_embd_params
         self.dense_bias = dense_bias
         self.q_scaling = q_scaling
+        self.attn_output_gate = attn_output_gate
+
+        if self.attn_output_gate:
+            logger.warning_once("using attn output gate!",
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py:170]: Why we always need warning when `attn_output_gate` enabled? Can we use `logger.info_once` instead?

File: tensorrt_llm/_torch/modules/attention.py [hunk 4]
```diff
@@ -207,7 +214,8 @@ def __init__(
 
         self.qkv_proj = Linear(
             self.hidden_size,
-            tp_size * self.q_size + 2 * tp_size * self.kv_size,
+            tp_size * self.q_size * (1 + (1 if self.attn_output_gate else 0)) +
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py:217]: Can we use `2 if self.attn_output_gate else 1` instead?

File: tensorrt_llm/_torch/modules/attention.py [hunk 5]
```diff
@@ -551,7 +574,8 @@ def apply_rope(self, q: torch.Tensor, k: Optional[torch.Tensor],
         """
         # If RoPE is fused into the attention OP, do not apply RoPE here.
         if not self.rope_fusion and position_ids is not None:
-            q, k, v = self.split_qkv(q, k, v)
+            if k is None and v is None:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py:577]: Don't need to check k/v, we have checked them in `split_qkv`.

File: tensorrt_llm/_torch/modules/rms_norm.py [hunk 6]
```diff
@@ -29,25 +29,31 @@ class RMSNorm(nn.Module):
...
-        device: Optional[torch.device] = None,
-        has_weights: bool = True,
+            self,
+            *,
+            hidden_size: int,
+            eps: float,
+            dtype: Optional[torch.dtype] = None,
+            device: Optional[torch.device] = None,
+            has_weights: bool = True,
+            use_gemma_rms_norm: bool = False,  # Assume has_weights = True
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/rms_norm.py:39]: Use assertion instead of comment. We can add `assert has_weights, "xxx"` when `use_gemma_rms_norm` is True.

Comment (yuxianq) [inline tensorrt_llm/_torch/modules/rms_norm.py:39]: Don't need the `_rms_norm` suffix since we already know it is a `RMSNorm` module. How about `use_gemma` or `use_gemma_style`?

File: tensorrt_llm/_torch/modules/qk_norm_attention.py [hunk 7]
```diff
@@ -156,11 +156,20 @@ def __init__(
...
         self.pretrained_config = config.pretrained_config
 
         self.fuse_qk_norm_rope = fuse_qk_norm_rope
         self.skip_rope = skip_rope
+        if use_gemma_rms_norm:
+            assert fuse_qk_norm_rope is False, "fused_qk_norm_rope is not supported for gemma rms norm."
+
+        # If fuse_qk_norm_rope is true, do not apply fused RoPE in attention OP, and self.rotary_emb will be skipped in the overridden apply_rope.
+        rope_fusion = not self.fuse_qk_norm_rope and not skip_rope
+        if attn_output_gate and use_gemma_rms_norm:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/qk_norm_attention.py:171]: Why `attn_output_gate and use_gemma_rms_norm` instead of `attn_output_gate or use_gemma_rms_norm`? Does we support rope fusion when one of them is enabled?

Comment (byshiue) [reply tensorrt_llm/_torch/modules/qk_norm_attention.py:171]: `attn_output_gate and use_gemma_rms_norm` means the qwen3-next model to prevent affecting other models. But I think it is fine to change it to `attn_output_gate or use_gemma_rms_norm` now if guoming does not encounter any functionality issue on qwen3-next and CI

File: tensorrt_llm/_torch/modules/attention.py [hunk 8]
```diff
@@ -511,24 +519,39 @@ def forward(
             if qkv_lora is not None:
                 qkv = qkv + qkv_lora
 
+        if self.attn_output_gate:
+            q_gate, k, v = qkv.split(
+                [self.q_size * 2, self.kv_size, self.kv_size], dim=-1)
```
[the remark this answers is not in the document]
Comment (byshiue) [reply tensorrt_llm/_torch/modules/attention.py:524]: I guess it is fine. This is referred from the other implementation.

File: tensorrt_llm/_torch/modules/qk_norm_attention.py (hunk 7 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/modules/qk_norm_attention.py:171]: Then we can use the simpler `rope_fusion = not self.fuse_qk_norm_rope and not skip_rope and not attn_output_gate and not use_gemma_rms_norm` instead.

File: tensorrt_llm/_torch/modules/attention.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/modules/attention.py:524]: Got it.

File: cpp/tensorrt_llm/kernels/fusedQKNormRopeKernel.cu
Comment (RobertLou) [inline cpp/tensorrt_llm/kernels/fusedQKNormRopeKernel.cu]: Qwen-next Head dim is 256，Add implement

File: tensorrt_llm/_torch/models/modeling_qwen3.py
Comment (RobertLou) [inline tensorrt_llm/_torch/models/modeling_qwen3.py]: 修改qwen3的接口信息，对于config.json补充的值使用getattr启用默认值
